### ISSUE TITLE
jupyter-labhub --collaborative does not work with docker spawner.

### ISSUE THREAD

**Issue by efagerberg** [jupyterlab/jupyterlab#11394]
## Description

When UserA shares a notebook with UserB and the jupyterhub instance is using the dockerspawner along with `--collaborative`, UserB is not able to collaborate with UserA on that notebook/

## Reproduce

1. Clone https://github.com/efagerberg/jupyterlab_collab_bug_dockerspawner
2. docker-compose build
3. docker-compose up -d jupyterhub
4. in a browser incognito tab, go to 127.0.0.1:8000 and login as UserA:UserA
5. in a separate browser incognito tab (needs to be a different browser to enforce no cookie sharing) go to 127.0.0.1:8000 and login as UserB:UserB
6. As UserA, create a notebook in the root directory in jupyterhub.
7. Copy the sharable link to that notebook (right click on notebook, copy sharable link)
8. Paste sharable link in UserB session
9. Observe the error message stating that UserB cannot find the file in question.
10. As UserA, create a notebook in the shared directory.
11. As UserA, copy the sharable notebook link for the notebook in the shared directory.
12. Paste sharable link in UserB session
13. Observe, that UserB can access the file, but no live collaboration session starts.

## Expected behavior

UserA should be able to share a notebook with UserB and start a RTC session. Perhaps the first case is special enough to not be of interest (how can UserB access the notebook if it is outside of the context of their isolated container). But I would expect if they both have access to that file that the collaboration session would work.

## Context

- Operating System and version: Ubuntu 21.10
- Browser and version: Chrome Version 95.0.4638.69 (Official Build) (64-bit) Firefox: Version 95.0.4638.69 (Official Build) (64-bit) 
- JupyterLab version: 3.2.0

**Comment by hbcarlos** [comment]
Hi @efagerberg. The "copy sharable link" you mention is not part of the RTC. That button is for JupyterHub.

On the other side, are both users in the same JupyterLab instance, or is JupyterHub creating a new JupyterLab instance for each user?
To collaborate, both users need to be working on the same JupyterLab server; otherwise, they can not collaborate.

**Comment by efagerberg** [comment]
I see I did not realize that they needed to be the same instance. In this case dockerspawner creates a container with jupyterlab per user.

I suppose then the link share extension would be my next thing to test. Ideally that makes a link that just points to that user's session. Unfortunately in that example I am not able to install that extension, I always get a build error.

Is there another way to trigger a rtc session without that extension?

**Comment by hbcarlos** [comment]
Have you tried to copy the URL from UserA. Probably, you'll need to log in as UserA to access the first server but I'm not sure how JupyterHub works.

**Comment by efagerberg** [comment]
Yeah I actually got that to work once when I was trying to reproduce but did not realize 2 incognito windows in chrome were sharing the same browser state resulting in the same user in both windows.

Basically the flow is User logs in, jupyterhub uses it's notebook spawner in this case to spawn a docker container running `jupyter-labhub --collaborative`. I think ideally the copy sharable link action would do something different for collab notebooks in which it would include any needed tokens in the url to auth as the sharing user. At least that is how I would see it working with the way jupyterlab works right now.

Obviously this has some quirks to it. If UserB auths as that user they likely would stay authed as that user for some time, unless the link's tokens are single use only.

I think the ideal world solution is that a user in the jupyterhub instance would be able to share their notebook with UserB without UserB needing to technically be logged in as UserA, and without the requirement that the notebook live in some shared data volume between the two users, although i would understand holding off of that because it does seem docker specific.

**Comment by efagerberg** [comment]
https://discourse.jupyter.org/t/plans-on-bringing-rtc-to-jupyterhub/9813/3 has some interesting discussion on the topic.
